### ISSUE TITLE
LookAheadOptimizer导致模型不收敛

### ISSUE THREAD

**Issue by flishwang** [PaddlePaddle/PaddleDetection#1262]
我基于release/0.3的版本，在tools/train.py中增加如下内容：

                lr = lr_builder()
                optimizer = optim_builder(lr)
                if cfg.get('lookahead',-1.0) > 1e-3:
                    optimizer= LookAheadOptimizer(optimizer,alpha=cfg.get('lookahead',-1.0),k=5)
                    optimizer.minimize(loss,startup_program=startup_prog)
                else:
                    optimizer.minimize(loss)

模型使用大规模目标检测cascade_rcnn r101vd dcn的那个模型，
但模型训练后mAP只有0.01左右。

根据https://github.com/PaddlePaddle/Paddle/pull/25688 对Optimizer进行修改后，问题依旧没有得到解决。

将LookAheadOptimizer做下文所示的修改后，模型精度恢复，与不加lookahead相似，怀疑是optimizer中slow_params的初始化有问题。

我对lookaheadOptimizer的修改：

        with fluid.framework.program_guard(main_block.program,startup_program):
            # Add Var k to main prog and startup prog
            k = fluid.layers.create_global_var(
                name="lookahead_k",
                shape=[1],
                value=int(self.k),
                dtype='int32',
                persistable=True)

            # Add Var alpha to main prog and startup prog
            alpha = fluid.layers.create_global_var(
                name="lookahead_alpha",
                shape=[1],
                value=1.0,
                dtype='float32',
                persistable=True)

            alpha2 = fluid.layers.create_global_var(
               name="lookahead_alpha2",
               shape=[1],
               value=float(self.alpha),
               dtype='float32',
               persistable=True)   #这里增加alpha2, 把alpha的初值设为1.0

            # Add Var step
            step = fluid.layers.create_global_var(
                name="lookahead_step",
                shape=[1],
                value=int(0),
                dtype='int32',
                persistable=True)
            fluid.layers.increment(x=step, value=1.0, in_place=True)

            # lookahead
            zero_var = fluid.layers.fill_constant(shape=[1], dtype='float32', value=0.0)

            one_var = fluid.layers.fill_constant(shape=[1], dtype='float32', value=1.0)

            mod = fluid.layers.elementwise_mod(step, k)
            with fluid.layers.control_flow.Switch() as switch:
                with switch.case(step == k * 8):   #这里增加一个switch case, 在step=40时把alpha从1.0更改为self.alpha。k=2时似乎训练也没有问题。
                    fluid.layers.assign(input=alpha2, output=alpha)
                with switch.case(mod == zero_var):
                    for param_name in params:
                        fast_var = main_block.var(param_name)
                        slow_var = param_to_slow[param_name]

                        tmp_var = fluid.layers.elementwise_add(
                            fluid.layers.elementwise_mul(fast_var, alpha),
                            fluid.layers.elementwise_mul(slow_var, one_var - alpha))
                        fluid.layers.assign(input=tmp_var, output=slow_var)
                        fluid.layers.assign(input=tmp_var, output=fast_var)
                with switch.default():
                    pass

        return mini_out

**Comment by mapingshuo** [comment]
怀疑是slow_param数值有问题的话，可以打印下来看看；至于Lookahead收敛效果不够好，也是符合预期的，得具体分析

**Comment by flishwang** [comment]
> 怀疑是slow_param数值有问题的话，可以打印下来看看；

好的。我对代码进行了如下的更改：



            with fluid.layers.control_flow.Switch() as switch:
                with switch.case(step == one_var):
                    fluid.layers.Print(main_block.var('conv1_1_weights'))
                    fluid.layers.Print(param_to_slow['conv1_1_weights'])
                with switch.case(step == k * 8):
                    fluid.layers.assign(input=alpha2, output=alpha)
                with switch.case(mod == zero_var):
                    for param_name in params:
                        fast_var = main_block.var(param_name)
                        slow_var = param_to_slow[param_name]

                        tmp_var = fluid.layers.elementwise_add(
                            fluid.layers.elementwise_mul(fast_var, alpha),
                            fluid.layers.elementwise_mul(slow_var, one_var - alpha))
                        fluid.layers.assign(input=tmp_var, output=slow_var)
                        fluid.layers.assign(input=tmp_var, output=fast_var)
                with switch.default():
                    pass



打印日志如下：


2020-08-20 19:41:55,997-INFO: Total iters are 3568 under 1 devices
2020-08-20 19:41:55,998-INFO: set (<ppdet.optimizer.PiecewiseDecay object at 0x7fcc3636b850>).milestones to [5606, 9684]
2020-08-20 19:41:55,998-INFO: set base learning_rate to 0.004687500000000001
2020-08-20 19:42:00,121-INFO: If regularizer of a Parameter has been set by 'fluid.ParamAttr' or 'fluid.WeightNormParamAttr' already. The Regularization[L2Decay, regularization_coeff=0.000100] in Optimizer will not take effect, and it will only be applied to other Parameters!
W0820 19:42:02.596233  5080 device_context.cc:252] Please NOTE: device: 0, CUDA Capability: 75, Driver API Version: 10.1, Runtime API Version: 10.0
W0820 19:42:02.598673  5080 device_context.cc:260] device: 0, cuDNN Version: 7.6.
2020-08-20 19:42:06,095-INFO: places would be ommited when DataLoader is not iterable
1597923729		The place is:CUDAPlace(0)
Tensor[conv1_1_weights]
	shape: [32,3,3,3,]
	dtype: f
	data: 0.357185,0.494535,0.44597,0.011593,0.0198742,-0.0238901,-0.311222,-0.535815,-0.434869,0.485142,0.706262,0.629845,-0.0232923,0.0139924,-0.0172578,-0.47487,-0.7465,-0.597637,0.207547,0.326455,
1597923729		The place is:CUDAPlace(0)
Tensor[conv1_1_weights@SLOW]
	shape: [32,3,3,3,]
	dtype: f
	data: -0.231623,-0.0257222,-0.226582,0.181279,0.0611427,0.0654207,-0.212313,-0.333934,0.518324,0.275623,0.567527,0.239242,-0.0990483,0.370958,-0.149964,0.547846,-0.43046,-0.315304,-0.17557,-0.184417,
2020-08-20 19:42:22,586-INFO: Queue current    0.0199 added   1.769  0.848  2.002  0.610
2020-08-20 19:42:22,588-INFO: iter: 20, lr: 0.001270, 'loss_cls_0': '1.896026', 'loss_loc_0': '0.018166', 'loss_cls_1': '0.948649', 'loss_loc_1': '0.020418', 'loss_cls_2': '0.165094', 'loss_loc_2': '0.013431', 'loss_rpn_cls': '25.961288', 'loss_rpn_bbox': '0.630786', 'loss': '0.399599', 'bonus_mean': '0.795139', time: 0.786, eta: 0:46:27

两组tensor完全不同，应该是有地方出问题了。

> 至于Lookahead收敛效果不够好，也是符合预期的，得具体分析

我之前贴了我对Lookahead的更改，改完后收敛效果就好，不改就差，我觉得这里边有bug，所以才提的issue，至少我不觉得mAP从0.80多掉到0.01左右是符合lookahead预期的。

**Comment by mapingshuo** [comment]
我是说，收敛效果不完美是符合预期的，但发散是不符合预期的

**Comment by mapingshuo** [comment]
请问alpha的值是多少时会发散？

**Comment by mapingshuo** [comment]
试试运行一下我这个demo，看看Slow_param与quick_param是否一致，我们排除一下paddle本身的问题：

```
import paddle.fluid.core as core
import paddle.fluid as fluid
import paddle.fluid.optimizer as optimizer
import numpy as np

def sample_data():
   res = []
   for i in range(2):
       data = np.random.normal(size=(2,))
       label = np.random.randint(2, size=(1,))
       res.append((data, label))
   return res

x = fluid.layers.data(name='x', shape=[2], dtype='float32')
label = fluid.layers.data(name="label", shape=[1], dtype="int64")

# define net here
y = fluid.layers.fc(input=[x], size=2, act="softmax")
#y2 = fluid.layers.assign(input=y)
loss = fluid.layers.cross_entropy(input=y, label=label)
loss = fluid.layers.mean(x=loss)

# define backward and optimizer here

#fluid.backward.append_backward(loss=loss)
#optimizer = fluid.optimizer.SGD(learning_rate=0.01)
sgd = fluid.optimizer.SGD(learning_rate=0.01)
optimizer = optimizer.LookaheadOptimizer(sgd,
					    alpha=0.5,
					    k=5)
optimizer.minimize(loss)

with open("main_program", "w") as f:
    f.write(str(fluid.default_main_program()))

with open("startup_program", "w") as f:
    f.write(str(fluid.default_startup_program()))

main_program = fluid.default_main_program()

place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())

feeder = fluid.DataFeeder(feed_list=[x, label], place=place)
scope = fluid.global_scope()
step = 0
while(step < 10):
    print('---------------------')
    step += 1
    print("alpha: ", scope.var("lookahead_alpha").get_tensor().__array__())
    print("k: ", scope.var("lookahead_k").get_tensor().__array__())
    print("step: ", scope.var("lookahead_step").get_tensor().__array__())
    print(scope.var("lookahead_debug").get_tensor().__array__())
    print(scope.var("fc_0.w_0").get_tensor().__array__())
    print(scope.var("fc_0.w_0@SLOW").get_tensor().__array__())
    batch_data = sample_data()
    exe.run(fluid.default_main_program(),
	    feed=feeder.feed(batch_data))
```

**Comment by flishwang** [comment]
> 请问alpha的值是多少时会发散？

alpha=0.5或0.7时确认会发散。

我觉得应该就是slow_param的初值的问题。原因：
1. 在我上文的回复中，我在iter=1时打印了conv1_1_weights和conv1_1_weights@SLOW（见回复末尾），您可以看到，它们的值差别太大，一个iter的训练不可能有这样的结果。

2. 在最初提问时，我也说明了，如果在iter<40时使用alpha=1.0，在iter > 40时使用新的alpha，就不会有模型发散的问题。

另外，我试了您的demo，没有问题。

Tensor[conv1_1_weights]
shape: [32,3,3,3,]
dtype: f
data: 0.357185,0.494535,0.44597,0.011593,0.0198742,-0.0238901,-0.311222,-0.535815,-0.434869,0.485142,0.706262,0.629845,-0.0232923,0.0139924,-0.0172578,-0.47487,-0.7465,-0.597637,0.207547,0.326455,
1597923729 The place is:CUDAPlace(0)
Tensor[conv1_1_weights@SLOW]
shape: [32,3,3,3,]
dtype: f
data: -0.231623,-0.0257222,-0.226582,0.181279,0.0611427,0.0654207,-0.212313,-0.333934,0.518324,0.275623,0.567527,0.239242,-0.0990483,0.370958,-0.149964,0.547846,-0.43046,-0.315304,-0.17557,-0.184417,

**Comment by flishwang** [comment]
> 试试运行一下我这个demo，看看Slow_param与quick_param是否一致，我们排除一下paddle本身的问题：
> 
> ```
> import paddle.fluid.core as core
> import paddle.fluid as fluid
> import paddle.fluid.optimizer as optimizer
> import numpy as np
> 
> def sample_data():
>    res = []
>    for i in range(2):
>        data = np.random.normal(size=(2,))
>        label = np.random.randint(2, size=(1,))
>        res.append((data, label))
>    return res
> 
> x = fluid.layers.data(name='x', shape=[2], dtype='float32')
> label = fluid.layers.data(name="label", shape=[1], dtype="int64")
> 
> # define net here
> y = fluid.layers.fc(input=[x], size=2, act="softmax")
> #y2 = fluid.layers.assign(input=y)
> loss = fluid.layers.cross_entropy(input=y, label=label)
> loss = fluid.layers.mean(x=loss)
> 
> # define backward and optimizer here
> 
> #fluid.backward.append_backward(loss=loss)
> #optimizer = fluid.optimizer.SGD(learning_rate=0.01)
> sgd = fluid.optimizer.SGD(learning_rate=0.01)
> optimizer = optimizer.LookaheadOptimizer(sgd,
> 					    alpha=0.5,
> 					    k=5)
> optimizer.minimize(loss)
> 
> with open("main_program", "w") as f:
>     f.write(str(fluid.default_main_program()))
> 
> with open("startup_program", "w") as f:
>     f.write(str(fluid.default_startup_program()))
> 
> main_program = fluid.default_main_program()
> 
> place = fluid.CPUPlace()
> exe = fluid.Executor(place)
> exe.run(fluid.default_startup_program())
> 
> feeder = fluid.DataFeeder(feed_list=[x, label], place=place)
> scope = fluid.global_scope()
> step = 0
> while(step < 10):
>     print('---------------------')
>     step += 1
>     print("alpha: ", scope.var("lookahead_alpha").get_tensor().__array__())
>     print("k: ", scope.var("lookahead_k").get_tensor().__array__())
>     print("step: ", scope.var("lookahead_step").get_tensor().__array__())
>     print(scope.var("lookahead_debug").get_tensor().__array__())
>     print(scope.var("fc_0.w_0").get_tensor().__array__())
>     print(scope.var("fc_0.w_0@SLOW").get_tensor().__array__())
>     batch_data = sample_data()
>     exe.run(fluid.default_main_program(),
> 	    feed=feeder.feed(batch_data))
> ```

有没有可能是这段代码的问题？
https://github.com/PaddlePaddle/PaddleDetection/blob/release/0.4/tools/train.py，179行：


    exe.run(startup_prog)


194-201行：
    if FLAGS.resume_checkpoint:
        checkpoint.load_checkpoint(exe, train_prog, FLAGS.resume_checkpoint)
        start_iter = checkpoint.global_step()
    elif cfg.pretrain_weights and fuse_bn and not ignore_params:
        checkpoint.load_and_fusebn(exe, train_prog, cfg.pretrain_weights)
    elif cfg.pretrain_weights:
        checkpoint.load_params(
            exe, train_prog, cfg.pretrain_weights, ignore_params=ignore_params)

导致在startup_prog里，param和slow_param是同步的，但param从checkpoint.load_params里获取了预训练模型的值，但slow_params未获取，因此出的问题？

**Comment by mapingshuo** [comment]
你说得对，是因为load_params更新了模型参数，但是没有更新slow_params导致

**Comment by mapingshuo** [comment]
你看一下修复是否能解决问题：https://github.com/PaddlePaddle/Paddle/compare/develop...mapingshuo:lookahead_initialize?expand=1

**Comment by mapingshuo** [comment]
你看以下修复能否解决你的问题：https://github.com/PaddlePaddle/Paddle/pull/26516

**Comment by flishwang** [comment]
> 你看以下修复能否解决你的问题：[PaddlePaddle/Paddle#26516](https://github.com/PaddlePaddle/Paddle/pull/26516)

这样改可以的。谢谢！

**Comment by mapingshuo** [comment]
客气了，感谢您的反馈，合作愉快
